File: GoogleNavigation/GoogleNavXCFrameworkDemos/Samples/StopoverViewController.m
/*
 * Copyright 2020 Google LLC. All rights reserved.
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 * ANY KIND, either express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

#import "GoogleNavXCFrameworkDemos/Samples/StopoverViewController.h"

#import <CoreLocation/CoreLocation.h>

#if __has_feature(modules)
@import GoogleMaps;
#else
#import <GoogleMaps/GoogleMaps.h>
#endif
#if __has_feature(modules)
@import GoogleNavigation;
#else
#import <GoogleNavigation/GoogleNavigation.h>
#endif
#import "GoogleNavXCFrameworkDemos/Utils/NavDemoStringUtils.h"
#import "GoogleNavXCFrameworkDemos/Utils/NavDemoSwitch.h"
#import "GoogleNavXCFrameworkDemos/Utils/NavDemoUtilities.h"

/** The origin coordinations of demos. */
static const CLLocationCoordinate2D kSingaporeTunnelOriginCoord1 = {1.295720, 103.848683};
static const CLLocationCoordinate2D kSingaporeTunnelOriginCoord2 = {1.298818, 103.877174};
static const CLLocationCoordinate2D kJarkadaHighwayOriginCoord = {-6.126155, 106.849174};
static const CLLocationCoordinate2D kManilaHighwayOriginCoord = {14.502297, 121.0351525};

/** The destination coordinations of demos. */
static const CLLocationCoordinate2D kSingaporeTunnelDestCoord1 = {1.297535, 103.846630};
static const CLLocationCoordinate2D kSingaporeTunnelDestCoord2 = {1.302525, 103.878126};
static const CLLocationCoordinate2D kJarkadaHighwayDestCoord = {-6.122432, 106.859372};
static const CLLocationCoordinate2D kManilaHighwayDestCoord = {14.503486, 121.036674};

@implementation StopoverViewController {
  BOOL _vehicleStopover;
  NSInteger _travelID;
  GMSMapView *_mapView;
}

- (void)viewDidLoad {
  [super viewDidLoad];
  _mapView = [[GMSMapView alloc] init];
  _mapView.navigationEnabled = YES;
  _mapView.cameraMode = GMSNavigationCameraModeFollowing;
  _mapView.travelMode = GMSNavigationTravelModeDriving;
  [self.mainStackView addArrangedSubview:_mapView];
  GMSCameraPosition *camera = [GMSCameraPosition cameraWithLatitude:1.295720
                                                          longitude:103.848683
                                                               zoom:13];
  UIStackView *controls = self.controls;
  _mapView.camera = camera;
  _mapView.settings.recenterButtonEnabled = YES;
  [_mapView.locationSimulator simulateLocationAtCoordinate:kSingaporeTunnelOriginCoord1];
  [self.mainStackView addArrangedSubview:controls];

  // Add a button to request the route.
  UIButton *requestRouteButton =
      GMSNavigationCreateButton(self, @selector(requestRoute), @"Request route");
  [controls addArrangedSubview:requestRouteButton];

  // Add a button to clear the destination.
  UIButton *clearDestinationButton =
      GMSNavigationCreateButton(self, @selector(clearDestination), @"Clear destination");
  [controls addArrangedSubview:clearDestinationButton];

  // Add a stopover switch.
  UIView *stopoverSwitch = [NavDemoSwitch switchWithLabel:@"Vehicle Stopover"
                                                   target:self
                                                 selector:@selector(updateVehicleStopover:)];
  [controls addArrangedSubview:stopoverSwitch];

  // Add segmented controls to Travel ID.
  UISegmentedControl *travelIDControl = GMSNavigationCreateSegmentedControl(
      self, @selector(updateTravelID:), @[ @"Tunnel 1", @"Tunnel 2", @"Freeway 1", @"Freeway 2" ]);
  [controls addArrangedSubview:travelIDControl];
}

- (void)requestRoute {
  NSArray<GMSNavigationWaypoint *> *waypoints = [self createDestinationWaypoints];
  [_mapView.navigator clearDestinations];
  __weak StopoverViewController *weakSelf = self;
  GMSRouteStatusCallback callback = ^(GMSRouteStatus routeStatus) {
    [weakSelf handleRouteCallbackWithStatus:routeStatus];
  };
  [_mapView.navigator setDestinations:waypoints callback:callback];
}

/** Returns a list of destination waypoints based on the currently selected destination type. */
- (NSArray<GMSNavigationWaypoint *> *)createDestinationWaypoints {
  NSMutableArray<GMSNavigationMutableWaypoint *> *waypoints = [NSMutableArray array];
  switch (_travelID) {
    case 0: {
      GMSNavigationMutableWaypoint *waypoint =
          [[GMSNavigationMutableWaypoint alloc] initWithLocation:kSingaporeTunnelDestCoord1
                                                           title:@"Case 1 destination"];
      if (waypoint) {
        waypoint.vehicleStopover = _vehicleStopover;
        [waypoints addObject:waypoint];
      }
    } break;
    case 1: {
      GMSNavigationMutableWaypoint *waypoint =
          [[GMSNavigationMutableWaypoint alloc] initWithLocation:kSingaporeTunnelDestCoord2
                                                           title:@"Case 2 destination"];
      if (waypoint) {
        waypoint.vehicleStopover = _vehicleStopover;
        [waypoints addObject:waypoint];
      }
    } break;
    case 2: {
      GMSNavigationMutableWaypoint *waypoint =
          [[GMSNavigationMutableWaypoint alloc] initWithLocation:kJarkadaHighwayDestCoord
                                                           title:@"Case 3 destination"];
      if (waypoint) {
        waypoint.vehicleStopover = _vehicleStopover;
        [waypoints addObject:waypoint];
      }
    } break;
    case 3: {
      GMSNavigationMutableWaypoint *waypoint =
          [[GMSNavigationMutableWaypoint alloc] initWithLocation:kManilaHighwayDestCoord
                                                           title:@"Case 4 destination"];
      if (waypoint) {
        waypoint.vehicleStopover = _vehicleStopover;
        [waypoints addObject:waypoint];
      }
    } break;
  }
  return waypoints;
}

- (void)handleRouteCallbackWithStatus:(GMSRouteStatus)routeStatus {
  if (routeStatus != GMSRouteStatusOK) {
    // Show an error dialog to describe the failure.
    GMSNavigationPresentAlertController(self, GMSNavigationDemoMessageForRouteStatus(routeStatus),
                                        @"Route failed", @"OK");
  }
}

- (void)clearDestination {
  _mapView.navigator.guidanceActive = NO;
  [_mapView.navigator clearDestinations];
}

- (void)updateVehicleStopover:(NavDemoSwitch *)sender {
  _vehicleStopover = sender.on;
}

- (void)updateTravelID:(UISegmentedControl *)sender {
  _travelID = sender.selectedSegmentIndex;
  [self simulateToStartLocation];
}

- (void)simulateToStartLocation {
  CLLocationCoordinate2D coordinate = kCLLocationCoordinate2DInvalid;
  switch (_travelID) {
    case 0:
      coordinate = kSingaporeTunnelOriginCoord1;
      break;
    case 1:
      coordinate = kSingaporeTunnelOriginCoord2;
      break;
    case 2:
      coordinate = kJarkadaHighwayOriginCoord;
      break;
    case 3:
      coordinate = kManilaHighwayOriginCoord;
      break;
  }

  if (CLLocationCoordinate2DIsValid(coordinate)) {
    [_mapView.locationSimulator simulateLocationAtCoordinate:coordinate];
  }
}
@end

File: GoogleNavigation-Swift/GoogleNavSwiftXCFrameworkDemos/Swift/Samples/StopoverViewController.swift
/// Copyright 2020 Google LLC. All rights reserved.
///
///
/// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
/// file except in compliance with the License. You may obtain a copy of the License at
///
///     http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing, software distributed under
/// the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
/// ANY KIND, either express or implied. See the License for the specific language governing
/// permissions and limitations under the License.

import GoogleNavigation
import UIKit

class StopoverViewController: BaseSampleViewController {

  /// Hard-coded example destination waypoints that can be used to demonstrate how the "vehcile
  /// stopover" flag affects routes.
  private enum DestinationWaypoint: CaseIterable, CustomStringConvertible {
    case tunnel1
    case tunnel2
    case freeway1
    case freeway2

    var description: String {
      switch self {
      case .tunnel1:
        return "Tunnel 1"
      case .tunnel2:
        return "Tunnel 2"
      case .freeway1:
        return "Freeway 1"
      case .freeway2:
        return "Freeway 2"
      }
    }

    var waypoints: [GMSNavigationMutableWaypoint] {
      let title = "Stopover destination"
      var waypoints: [GMSNavigationMutableWaypoint] = []
      switch self {
      case .tunnel1:
        if let waypoint = GMSNavigationMutableWaypoint(
          location: CLLocationCoordinate2D(latitude: 1.297535, longitude: 103.846630),
          title: title)
        {
          waypoints.append(waypoint)
        }
      case .tunnel2:
        if let waypoint = GMSNavigationMutableWaypoint(
          location: CLLocationCoordinate2D(latitude: 1.302525, longitude: 103.878126),
          title: title)
        {
          waypoints.append(waypoint)
        }
      case .freeway1:
        if let waypoint = GMSNavigationMutableWaypoint(
          location: CLLocationCoordinate2D(latitude: -6.122432, longitude: 106.859372),
          title: title)
        {
          waypoints.append(waypoint)
        }
      case .freeway2:
        if let waypoint = GMSNavigationMutableWaypoint(
          location: CLLocationCoordinate2D(latitude: 14.503486, longitude: 121.036674),
          title: title)
        {
          waypoints.append(waypoint)
        }
      }
      return waypoints
    }

    var startCoordinate: CLLocationCoordinate2D {
      switch self {
      case .tunnel1:
        return CLLocationCoordinate2D(latitude: 1.295720, longitude: 103.848683)
      case .tunnel2:
        return CLLocationCoordinate2D(latitude: 1.298818, longitude: 103.877174)
      case .freeway1:
        return CLLocationCoordinate2D(latitude: -6.126155, longitude: 106.849174)
      case .freeway2:
        return CLLocationCoordinate2D(latitude: 14.502297, longitude: 121.0351525)
      }
    }
  }
  private let options: [DestinationWaypoint] = DestinationWaypoint.allCases
  private var vehicleStopover = false

  private lazy var mapView: GMSMapView = {
    let camera = GMSCameraPosition(latitude: 1.295720, longitude: 103.848683, zoom: 13)
    let options = GMSMapViewOptions()
    options.camera = camera
    options.frame = .zero
    let mapView = GMSMapView(options: options)
    mapView.isNavigationEnabled = true
    mapView.cameraMode = .following
    mapView.settings.isRecenterButtonEnabled = true
    mapView.travelMode = .driving
    return mapView
  }()

  private lazy var travelIDControl: UISegmentedControl = {
    let segmentedControl = UISegmentedControl(items: options.map { "\($0)" })
    segmentedControl.addTarget(self, action: #selector(simulateToStartLocation), for: .valueChanged)
    segmentedControl.selectedSegmentIndex = 0
    return segmentedControl
  }()

  override func viewDidLoad() {
    super.viewDidLoad()
    view.backgroundColor = .black

    // Add a map.
    primaryStackView.insertArrangedSubview(mapView, at: 0)
    mapView.locationSimulator?.simulateLocation(at: DestinationWaypoint.tunnel1.startCoordinate)

    // Add a button to request the route.
    addMenuSubview(
      MenuUIHelpers.makeMenuButton(
        title: "Request route", onTouchUpInside: (target: self, action: #selector(requestRoute))))

    // Add a button to clear the destination.
    addMenuSubview(
      MenuUIHelpers.makeMenuButton(
        title: "Clear destination",
        onTouchUpInside: (target: self, action: #selector(clearDestination))))

    // Add a stopover switch.
    addMenuSubview(
      NavDemoSwitch(
        title: "Vehicle Stopover",
        onValueChanged: (target: self, action: #selector(updateVehicleStopover))))

    // Add segmented controls to Travel ID.
    addMenuSubview(travelIDControl)
  }

  @objc func requestRoute() {
    let waypoints: [GMSNavigationMutableWaypoint] = options[travelIDControl.selectedSegmentIndex]
      .waypoints
    waypoints.forEach { (waypoint) in
      waypoint.vehicleStopover = vehicleStopover
    }
    mapView.navigator?.clearDestinations()
    mapView.navigator?.setDestinations(waypoints) { [weak self] (routeStatus) in
      self?.handleRouteCallback(with: routeStatus)
    }
  }

  /// Handles a route response with the given success or failure status.
  private func handleRouteCallback(with routeStatus: GMSRouteStatus) {
    guard routeStatus != .OK else { return }
    // Show an error dialog to describe the failure.
    let message = routeStatus.description
    let alertController = UIAlertController(
      title: "Route failed", message: message, preferredStyle: .alert)
    let defaultAction = UIAlertAction(title: "OK", style: .default)
    alertController.addAction(defaultAction)
    present(alertController, animated: true)
  }

  @objc func clearDestination() {
    guard let navigator = mapView.navigator else { return }
    navigator.isGuidanceActive = false
    navigator.clearDestinations()
  }

  @objc func simulateToStartLocation() {
    let startCoordinate = options[travelIDControl.selectedSegmentIndex].startCoordinate
    if CLLocationCoordinate2DIsValid(startCoordinate) {
      mapView.locationSimulator?.simulateLocation(at: startCoordinate)
    }
  }

  @objc func updateVehicleStopover(sender: UISwitch) {
    vehicleStopover = sender.isOn
  }
}
